src/Functions/printf.cpp

```diff
@@ -0,0 +1,309 @@
+#include <Columns/ColumnFixedString.h>
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnStringHelpers.h>
+#include <DataTypes/DataTypeString.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <Functions/formatString.h>
+#include <IO/WriteHelpers.h>
+
+#include <memory>
+#include <sstream>
+#include <string>
+#include <vector>
+#include <fmt/format.h>
+#include <fmt/printf.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int ILLEGAL_COLUMN;
+extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+namespace
+{
+
+class FunctionPrintf : public IFunction
+{
+private:
+    ContextPtr context;
+    FunctionOverloadResolverPtr function_concat;
+
+    struct Instruction
+    {
+        std::string_view format;
+        size_t rows;
+        bool is_literal; /// format is literal string without any argument
+        ColumnWithTypeAndName input; /// Only used when is_literal is false
+
+        ColumnWithTypeAndName execute()
+        {
+            if (is_literal)
+                return executeLiteral(format);
+            else if (isColumnConst(*input.column))
+                return executeConstant(input);
+            else
+                return executeNonconstant(input);
+        }
+
+        String toString() const
+        {
+            std::ostringstream oss; // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+            oss << "format:" << format << ", rows:" << rows << ", is_literal:" << is_literal << ", input:" << input.dumpStructure()
+                << std::endl;
+            return oss.str();
+        }
+
+    private:
+        ColumnWithTypeAndName executeLiteral(std::string_view literal)
+        {
+            ColumnWithTypeAndName res;
+            auto str_col = ColumnString::create();
+            str_col->insert(fmt::sprintf(literal));
```

Comment:
In which cases we need execute `fmt::sprintf` for literal? Does string `"abc %% def"` considered as literal? 

```suggestion
            /// ...
            str_col->insert(fmt::sprintf(literal));
```

Reply:
if input format is '%%' without any arguments, then '%%' is treated as a literal. But it still need to by processed by `fmt::sprintf` to get the correct unescaped result, which is '%'.